swabbie-orca/src/main/kotlin/com/netflix/spinnaker/swabbie/orca/OrcaTaskMonitoringAgent.kt

```diff
@@ -0,0 +1,190 @@
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.swabbie.orca
+
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.config.SwabbieProperties
+import com.netflix.spinnaker.kork.core.RetrySupport
+import com.netflix.spinnaker.kork.eureka.RemoteStatusChangedEvent
+import com.netflix.spinnaker.swabbie.MetricsSupport
+import com.netflix.spinnaker.swabbie.repositories.TaskCompleteEventInfo
+import com.netflix.spinnaker.swabbie.repositories.TaskTrackingRepository
+import com.netflix.spinnaker.swabbie.discovery.DiscoveryActivated
+import com.netflix.spinnaker.swabbie.events.*
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import rx.Scheduler
+import rx.schedulers.Schedulers
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+import java.time.temporal.Temporal
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+import javax.annotation.PostConstruct
+import javax.annotation.PreDestroy
+
+@Component
+class OrcaTaskMonitoringAgent (
+  private val clock: Clock,
+  val registry: Registry,
+  private val swabbieProperties: SwabbieProperties,
+  private val taskTrackingRepository: TaskTrackingRepository,
+  private val orcaService: OrcaService,
+  private val applicationEventPublisher: ApplicationEventPublisher,
+  private val retrySupport: RetrySupport
+  ) : DiscoveryActivated, MetricsSupport(registry) {
+
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+  private val worker: Scheduler.Worker = Schedulers.io().createWorker()
+
+  private val timeoutMillis: Long = 5000
+  private val maxAttempts: Int = 3
+
+  override val onDiscoveryUpCallback: (event: RemoteStatusChangedEvent) -> Unit
+    get() = {
+      worker.schedulePeriodically({
+       monitorOrcaTasks()
+      }, getAgentDelay(), getAgentFrequency(), TimeUnit.SECONDS)
+    }
+
+  override val onDiscoveryDownCallback: (event: RemoteStatusChangedEvent) -> Unit
+    get() = { stop() }
+
+  @PostConstruct
+  private fun init() {
+    log.info("Initializing agent ${javaClass.simpleName}")
+    registry.gauge(lastRunAgeId.withTag("agentName", javaClass.simpleName), this) {
+      Duration
+        .between(it.getLastAgentRun(), clock.instant())
+        .toMillis().toDouble()
+    }
+  }
+
+  @PreDestroy
+  private fun stop() {
+    log.info("Stopping agent ${javaClass.simpleName}")
+    if (!worker.isUnsubscribed) {
+      worker.unsubscribe()
+    }
+  }
+
+  private fun monitorOrcaTasks() {
+    initialize()
+
+    val inProgressTasks = taskTrackingRepository.getInProgress()
+    if (inProgressTasks.isEmpty()) {
+      log.debug("No active orca tasks to monitor from ${javaClass.simpleName}.")
+    }
+
+    inProgressTasks
```

Comment:
When there are orca tasks to monitor, this agent monitors them (checks status) every 10 seconds (configurable), and emits the appropriate event when they complete. This is the substitute for `ReceiveChannel`s. 

There should be locking around reading from redis in - I'll add that in.

Reply:
Added locking.